### PR TITLE
Make files module compatible post closure compiler

### PR DESCRIPTION
**Pull request by prateekbh**
- After closure compiler is done, we re parse all the `.js` files in `dist` folder and look for ternary conditions with results of form `window.global: this` and replace `this` with `self`.

- Parsing done with `@babel/parser`
- Changing contents with `MagicString` to update Sourcemaps later

### CODE REVIEW

File: build-system/tasks/create-module-compatible-scripts/index.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * limitations under the License.
+ */
+
+const $$ = require('gulp-load-plugins')();
+const colors = require('ansi-colors');
+const gulp = $$.help(require('gulp'));
+const log = require('fancy-log');
+const MagicString = require('magic-string');
+const parser = require('@babel/parser');
+const through = require('through2');
```
Comment (prateekbh) [inline build-system/tasks/create-module-compatible-scripts/index.js]: find an existing dependency alternative

File: build-system/tasks/create-module-compatible-scripts/index.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    enter(path) {
+      if (path.node.type !== 'ConditionalExpression')
+      {return;}
+      const {node} = path;
+      const {consequent, alternate} = node;
+      if (consequent.type !== 'MemberExpression') {return;}
+      if (alternate.type !== 'ThisExpression') {return;}
+      const {object, property} = consequent;
+      if (object.name !== 'window' && property.name !== 'global') {return;}
+      magicString.overwrite(alternate.start, alternate.end, 'self');
```
Comment (kristoferbaxter) [inline build-system/tasks/create-module-compatible-scripts/index.js]: A nice pattern for these kind of changes is to collect the changes you'd like to apply, then after you've completed analysis and think the changes are safe... apply them all at once.

This lets you easily back out of applying changes if you determine something later then an earlier mutation would have had knowledge to.

Alternatively, you can also just return the original code when you bail out and keep this approach.

File: build-system/tasks/create-module-compatible-scripts/index.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+const colors = require('ansi-colors');
+const gulp = $$.help(require('gulp'));
+const log = require('fancy-log');
+const MagicString = require('magic-string');
+const parser = require('@babel/parser');
+const through = require('through2');
+const traverse = require('@babel/traverse').default;
+
+
+function transform(file, encoding, callback) {
```
Comment (kristoferbaxter) [inline build-system/tasks/create-module-compatible-scripts/index.js]: Can we add a JSDoc?

File: build-system/tasks/create-module-compatible-scripts/index.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  const code = file.contents.toString('utf8');
+  const ast = parser.parse(code);
```
Comment (kristoferbaxter) [inline build-system/tasks/create-module-compatible-scripts/index.js]: Let's verify that the parser doesn't need additional configuration when called given the input may be es2015+ syntax.

File: build-system/tasks/create-module-compatible-scripts/index.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ * https://github.com/google/closure-compiler/blob/36f332788d54803c3c1afe06a9d84bf4b9f4945b/src/com/google/javascript/jscomp/js/util/global.js#L44
+ * Read more here: http://exploringjs.com/es6/ch_modules.html#_browsers-scripts-versus-modules
+ *
+ *
+ * Changes `global?global:VARNAME}(this)` to `global?global:VARNAME}(self)`
+ */
+exports.createModuleCompatibleBundle = function(srcGlob, destFolder) {
+  return new Promise(resolve => {
+    const {green} = colors;
+    log(green('Starting babel process, post closure compiler'));
```
Comment (kristoferbaxter) [inline build-system/tasks/create-module-compatible-scripts/index.js]: Can we make this description a bit more targeted toward this exact transform?

File: package.json [hunk 6]
```diff
@@ -139,6 +141,7 @@
     "karma-super-dots-reporter": "0.2.0",
     "lazypipe": "1.0.1",
     "lolex": "3.0.0",
+    "magic-string": "^0.25.1",
```
Comment (kristoferbaxter) [inline package.json]: Let's not use the hat version here, can we specify the exact pinned version (`0.25.1`).

File: package.json [hunk 7]
```diff
@@ -37,10 +37,12 @@
   "devDependencies": {
     "@babel/core": "7.1.6",
     "@babel/helper-plugin-test-runner": "7.1.0",
+    "@babel/parser": "^7.2.0",
     "@babel/plugin-proposal-class-properties": "7.1.0",
     "@babel/plugin-transform-runtime": "7.1.0",
     "@babel/preset-env": "7.1.6",
     "@babel/runtime-corejs2": "7.1.5",
+    "@babel/traverse": "^7.1.6",
```
Comment (kristoferbaxter) [inline package.json]: Let's not use the hat version here, can we specify the exact pinned version (7.1.6).

File: package.json [hunk 8]
```diff
@@ -37,10 +37,12 @@
   "devDependencies": {
     "@babel/core": "7.1.6",
     "@babel/helper-plugin-test-runner": "7.1.0",
+    "@babel/parser": "^7.2.0",
```
Comment (kristoferbaxter) [inline package.json]: Let's not use the hat version here, can we specify the exact pinned version (7.2.0).

File: package.json [hunk 9]
```diff
@@ -177,7 +180,7 @@
     "sleep-promise": "8.0.1",
     "tempy": "0.2.1",
     "text-table": "0.2.0",
-    "through2": "3.0.0",
+    "through2": "^3.0.0",
```
Comment (kristoferbaxter) [inline package.json]: Let's not use the hat version here, can we specify the exact pinned version (`3.0.0`).

File: src/service/url-replacements-impl.js [hunk 10]
```diff
@@ -296,8 +296,8 @@ export class GlobalVariableSource extends VariableSource {
       return this.getVariantsValue_(variants => {
         const variant = variants[/** @type {string} */(experiment)];
         user().assert(variant !== undefined,
-            'The value passed to VARIANT() is not a valid experiment name:' +
-                experiment);
+            'The value passed to VARIANT() is not a valid experiment name:%s',
```
Comment (kristoferbaxter) [inline src/service/url-replacements-impl.js]: These changes, while nice, appear unrelated. Intentional?

File: package.json (hunk 8 above)
Comment (prateekbh) [reply package.json]: done

File: package.json (hunk 7 above)
Comment (prateekbh) [reply package.json]: done

File: package.json (hunk 6 above)
Comment (prateekbh) [reply package.json]: done

File: package.json (hunk 9 above)
Comment (prateekbh) [reply package.json]: removed

File: src/service/url-replacements-impl.js (hunk 10 above)
Comment (prateekbh) [reply src/service/url-replacements-impl.js]: reverted

File: build-system/tasks/create-module-compatible-scripts/index.js (hunk 5 above)
Comment (prateekbh) [reply build-system/tasks/create-module-compatible-scripts/index.js]: done

File: build-system/tasks/create-module-compatible-scripts/index.js (hunk 2 above)
Comment (prateekbh) [reply build-system/tasks/create-module-compatible-scripts/index.js]: done

File: build-system/tasks/create-module-compatible-scripts/index.js (hunk 3 above)
Comment (prateekbh) [reply build-system/tasks/create-module-compatible-scripts/index.js]: done

File: build-system/tasks/create-module-compatible-scripts/index.js (hunk 1 above)
Comment (prateekbh) [reply build-system/tasks/create-module-compatible-scripts/index.js]: done

File: build-system/tasks/create-module-compatible-scripts/index.js (hunk 4 above)
Comment (prateekbh) [reply build-system/tasks/create-module-compatible-scripts/index.js]: I cant think of any options listed [here](https://babeljs.io/docs/en/next/babel-parser.html#options) to be really necessary, but would like @jridgewell to take another look

**Comment by jridgewell** [comment]
https://astexplorer.net/#/gist/d0b64fe339a0051d0052d30a797c3c40/70a7addf0183ab8a09ee52764f3a79c968d8673b

**Comment by prateekbh** [comment]
@jridgewell i have added a few more tests to be doubly sure of and have modified the code to your suggestion.

Nit: need to exempt fixtures from lint may be

File: build-system/tasks/create-module-compatible-scripts/index.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    const {green} = colors;
+    log(green(
+        'Starting post closure compiler transform to make scripts module safe.'
+    ));
+    gulp.src(srcGlob)
+        .pipe($$.sourcemaps.init({loadMaps: true}))
+        .pipe(transformTopLevelGlobalScope())
+        .pipe($$.sourcemaps.write('./'))
+        .pipe(gulp.dest(destFolder))
+        .on('end', resolve);
```
Comment (jridgewell) [inline build-system/tasks/create-module-compatible-scripts/index.js]: I believe `gulp.src()...` returns promises?

File: build-system/tasks/create-module-compatible-scripts/index.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+ * @param {function(Error, Object)} callback - Call this function (optionally with an
+ * error argument and data) when you are done processing the supplied chunk.
+ */
+function transform(file, encoding, callback) {
+  const code = file.contents.toString('utf8');
+  const ast = parser.parse(code);
+  const magicString = new MagicString(code);
+  const changes = [];
+
+  traverse(ast, {
```
Comment (jridgewell) [inline build-system/tasks/create-module-compatible-scripts/index.js]: I don't understand why this wasn't implemented as a standard babel plugin.

File: build-system/tasks/create-module-compatible-scripts/index.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    enter(path) {
+      // Find the closest encosing function that is not an arrow.
+      // Arrows inherit their parent's `this` context.
+      let parent = path;
+      const {node} = path;
+      if (node.type === 'ThisExpression') {
```
Comment (jridgewell) [inline build-system/tasks/create-module-compatible-scripts/index.js]: `path.isThisExpression()`

File: build-system/tasks/create-module-compatible-scripts/index.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+ * error argument and data) when you are done processing the supplied chunk.
+ */
+function transform(file, encoding, callback) {
+  const code = file.contents.toString('utf8');
+  const ast = parser.parse(code);
+  const magicString = new MagicString(code);
+  const changes = [];
+
+  traverse(ast, {
+    enter(path) {
```
Comment (jridgewell) [inline build-system/tasks/create-module-compatible-scripts/index.js]: I believe you can change this to `ThisExpression` and skip all other nodes.

File: build-system/tasks/create-module-compatible-scripts/index.js (hunk 12 above)
Comment (prateekbh) [reply build-system/tasks/create-module-compatible-scripts/index.js]: I am not 100% familiar with the terminology so I dont know what do you mean by that? 
Are you saying that we wont need the magic string if we just re-run this file through babel?

Comment (jridgewell) [reply build-system/tasks/create-module-compatible-scripts/index.js]: Yes. Use babel to mutate the AST.

File: build-system/tasks/create-module-compatible-scripts/index.js (hunk 11 above)
Comment (jridgewell) [reply build-system/tasks/create-module-compatible-scripts/index.js]: Alright, we need to hook up an `.on('error')` event that rejects the promise.

File: build-system/tasks/create-module-compatible-scripts/index.js (hunk 12 above)
Comment (prateekbh) [reply build-system/tasks/create-module-compatible-scripts/index.js]: Actually I agree, a standard babel plugin will keep the codebase more streamlined to read through

**Comment by prateekbh** [comment]
Hmm... looks like this is no more a problem. 
The hard coded `window.global?window.global:this` is not present anymore and the `v0.js` runs fine with `<script type=module>`. May due to a closure compiler update.

e.g.: https://jsbin.com/begetonuna/edit?html,console

This seemed to have changed between Jul16 and Jul 24th Cycle.

I'll investigate as in what changed during that cycle.

Also I'll probably close this and add a test to check if our binaries load fine with `<script type=module` in future as well.

**Comment by jridgewell** [comment]
Current output is `var la=function(a){return"undefined"!=typeof window&&window===a?a:"undefined"!=typeof global&&null!=global?global:a}(this);`

**Comment by jridgewell** [comment]
The synthetic file injected has changed meaningfully: https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/js/util/global.js

**Comment by prateekbh** [comment]
well actually this works because we also also introduced [compile-wrapper](https://github.com/ampproject/amphtml/commit/16c65fb819f9ed385080e3df2a6662bd4a1c393d#diff-95e858fb03f9044ffe992cfcf32cd7e2R21) on Jul 17th.

**Comment by nainar** [comment]
Closing this PR as inactive. Please reopen if needed.
